services/download/src/main/java/app/coronawarn/server/services/download/FifoMaxEntriesSet.java

```diff
@@ -0,0 +1,56 @@
+package app.coronawarn.server.services.download;
+
+import java.util.HashSet;
+import java.util.LinkedList;
+import java.util.Queue;
+import java.util.Set;
+
+/**
+ * Implements a size-limited {@link Set}-like {@link Queue}: Elements are only
+ * added if they are not already contained and if the number of elements exceeds
+ * the given limit, oldest elements are removed (FIFO).
+ *
+ * @param <E> the type of elements held in this collection.
+ * 
+ * @see Queue
+ * @see Set
+ */
+public class FifoMaxEntriesSet<E> {
+
+  private final Set<E> fast = new HashSet<>();
+  private final Queue<E> order = new LinkedList<>();
+  private final int maxSize;
+
+  public FifoMaxEntriesSet(int maxSize) {
+    this.maxSize = maxSize;
+  }
+
+  /**
+   * Adds the element to this set and removes elements until size <= maxSize.
+   * 
+   * @param e The element to add
+   * @return {@code true} if this set did not already contain the specified
+   *         element
+   */
+  public boolean add(E e) {
+    fast.add(e);
+    order.add(e);
+    while (fast.size() > maxSize) {
+      fast.remove(order.remove());
+    }
```

Comment:
@sorin-stefan-iovita-sap / @roesslerj 
sorry, but I still don't get it... let's assume the following: we have max size of 3 and batch-IDs A, B, C, D, A, B, C, D, ...
when `fast` reaches `maxSize`, it contains A,B,C,D - it will remove A as it is the first element in `order`...
So with next batch being again ID: A, we are stuck in a loop again, aren't we?

Reply:
Yes. But there is a trade-off: Either no maxSize and possibly introduce memory leak—or have maxSize and not detect loops that are bigger than maxSize. I am ok with either, we just can't have both and need to decide what we want...

Reply:
Resolved as discussed :-)